File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void solve(vector<bool>& mark, vector<int>& ds, vector<int>& nums,
               vector<vector<int>>& ans) {
        if (ds.size() == nums.size()) {
            ans.push_back(ds);
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            if (!mark[i]) {
                mark[i] = true;
                ds.push_back(nums[i]);
                solve(mark, ds, nums, ans);
                ds.pop_back();
                mark[i] = false;
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<bool> mark(n, false);
        vector<int> ds;
        vector<vector<int>> ans;
        solve(mark, ds, nums, ans);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    void solve(List<List<Integer>> res, List<Integer> ds, int[] nums, boolean vis[]){
        if(ds.size() == nums.length){
            res.add(new ArrayList<>(ds));
            return;
        }

        for(int i = 0; i < nums.length; i++){
            if(vis[i] == false){
                vis[i] = true;
                ds.add(nums[i]);
                solve(res, ds, nums, vis);
                ds.remove(ds.size() - 1);
                vis[i] = false;
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();

        int n = nums.length;
        boolean[] vis = new boolean[n];
        Arrays.fill(vis, false);
        
        solve(res, ds, nums, vis);

        return res;
    }
}